Add `Networkpolicy`s for `istio-system`

axel7born opened the pull request:
**How to categorize this PR?**

/area networking
/kind enhancement

**What this PR does / why we need it**:
Provide strict network policies for the istio-system so that it can only establish network connectivity to the endpoints that it is configured to reach inside the cluster and the seed api-server.
This is a second line of defence in case of a bug or misconfiguration of istio.

**Which issue(s) this PR fixes**:
None

**Special notes for your reviewer**:

**Release note**:

```other operator
Deploy network policies to namespace istio-system to only allow traffic to configured endpoints inside the cluster and the seed api-server.
```

axel7born committed:
Add networkpolicies for istio-system

axel7born committed:
Move networkpolicies to a dedicated file.

axel7born committed:
Add tests.

axel7born committed:
Let networkpolicy controller manage networkpolicy

allow-to-seed-apiserver in namespace istio-system.

DockToFuture commented on pkg/gardenlet/controller/networkpolicy/namespace_control.go:
```diff
@@ -79,7 +79,7 @@ func (r *namespaceReconciler) Reconcile(ctx context.Context, request reconcile.R
 	log = log.WithValues("networkPolicy", client.ObjectKeyFromObject(networkPolicy))
 
 	// if the namespace is not the Garden or a Shoot namespace - delete the existing NetworkPolicy
```
you can also adapt the comment.

DockToFuture reviewed (approved):
/lgtm

axel7born committed:
Update comment to match code changes.

ScheererJ commented on pkg/operation/botanist/component/istio/charts/istio/istio-istiod/templates/deployment.yaml:
```diff
@@ -18,7 +18,7 @@ spec:
     metadata:
       labels:
-{{ .Values.labels | toYaml | indent 8 }}
+{{ .Values.labels | toYaml | indent 8 }}networking.gardener.cloud/to-seed-apiserver: allowed
```
It looks a bit strange if everything is on the same line. Maybe move the new label before the template.

```suggestion
        networking.gardener.cloud/to-seed-apiserver: allowed
{{ .Values.labels | toYaml | indent 8 }}
```

ScheererJ replied:
Thanks for changing it.

ScheererJ commented on pkg/operation/botanist/component/istio/networkpolicies.go:
```diff
@@ -0,0 +1,511 @@
+			},
+		},
+		{
+			name: "allow-from-shoot-vpn",
```
Why is this network policy necessary?

axel7born replied:
I've seen in a tcpdump that istiod gets connections from shoot-vpn pod. If I remove the network policy, the seed doesn't get ready:
```
Message:               error during apply of object "networking.istio.io/v1alpha3/EnvoyFilter/istio-ingress/proxy-protocol": Put "https://api.d046309.core.internal.dev.k8s.ondemand.com:443/apis/networking.istio.io/v1alpha3/namespaces/istio-ingress/envoyfilters/proxy-protocol": context deadline exceeded
```

ScheererJ replied:
Is this due to an istio web hook? Can you please check whether a validating/mutating web hook is registered?

axel7born replied:
Yes, there is a validating webhook.

ScheererJ replied:
Ok, that explains it. This should probably be added as a comment or in the description. Otherwise, we will wonder about this next time we look at it...

ScheererJ commented on pkg/operation/botanist/component/istio/networkpolicies.go:
```diff
@@ -0,0 +1,511 @@
+										v1beta1constants.GardenRole: v1beta1constants.GardenRoleSystemComponent,
+									},
+								},
+								NamespaceSelector: &metav1.LabelSelector{},
```
`vpn-shoot` only runs in `kube-system` namespace. There is no need to allow any namespace.

ScheererJ commented on pkg/operation/botanist/component/istio/networkpolicies.go:
```diff
@@ -0,0 +1,511 @@
+										v1beta1constants.GardenRole: "monitoring",
+									},
+								},
+								NamespaceSelector: &metav1.LabelSelector{},
```
Aggregate prometheus only runs in `garden` namespace. Therefore, it makes sense to use a specific namespace here.

ScheererJ commented on pkg/operation/botanist/component/istio/networkpolicies.go:
```diff
@@ -0,0 +1,511 @@
+							}},
+							Ports: []networkingv1.NetworkPolicyPort{{
+								Protocol: protocolPtr(corev1.ProtocolTCP),
+								Port:     func(port string) *intstr.IntOrString { v := intstr.FromString(port); return &v }("metrics"),
```
Is a function here really necessary?

Would this not be sufficient?

```suggestion
								Port:     &intstr.FromString("metrics"),
```

axel7born replied:
That doesn't work here.
It's not possible to get the address of a string literal. It needs to be assigned to a variable.

ScheererJ replied:
Ok.

ScheererJ commented on pkg/operation/botanist/component/istio/networkpolicies.go:
```diff
@@ -0,0 +1,511 @@
+								PodSelector: &metav1.LabelSelector{
+									MatchLabels: map[string]string{
+										v1beta1constants.LabelApp:   "aggregate-prometheus",
+										v1beta1constants.GardenRole: "monitoring",
```
Are there any constants for the values?

axel7born replied:
Fixed for "monitoring", but there is really no constant for "aggregate-prometheus".

ScheererJ commented on pkg/operation/botanist/component/istio/networkpolicies.go:
```diff
@@ -0,0 +1,511 @@
+}
+
+func getIstioSystemNetworkPolicyTransformers(values IstioNetworkPolicyValues) []networkPolicyTransformer {
+	return []networkPolicyTransformer{
```
How is this working? There is a deny all policy for egress and ingress, but only ingress policies. I would have expected that istiod needs to reach out to the seed`s kube-apiserver, i.e. there would be a egress policy required.

axel7born replied:
Yes, that was a bit tricky, as it's not trivial to determine the endpoints for the api server. However, as there is already such a networkpolicy for garden namespace and shoot namespaces, I just added istio-system there:https://github.com/gardener/gardener/blob/4377d31f71570bd88b05752c02b04436ba9621d3/pkg/gardenlet/controller/networkpolicy/namespace_control.go#L82
This needs an added label to the istiod pods. 

But I didn't want to add a dns resolve here.

ScheererJ replied:
Not sure if a comment would be helpful here so that nobody is confused in the future.

ScheererJ reviewed (changes requested):
Please check the multi-az tests and clarify my questions.

axel7born committed:
Address requested changes:

* Add namespace selectors for networkpolicies.
* Change format in deployment chart.
* Use constant for string "monitoring"

axel7born commented:
/test pull-gardener-e2e-kind-ha-single-zone

ScheererJ commented on pkg/operation/botanist/component/istio/networkpolicies.go:
```diff
@@ -0,0 +1,519 @@
+							From: []networkingv1.NetworkPolicyPeer{{
+								PodSelector: &metav1.LabelSelector{
+									MatchLabels: map[string]string{
+										v1beta1constants.LabelApp:   "vpn-shoot",
```
Please use the constant from `vpnshoot.go`, which needs to be exposed first.

ialidzhikov commented on pkg/operation/botanist/component/istio/istiod_test.go:
```diff
@@ -1916,6 +1917,147 @@ spec:
+  annotations:
+    gardener.cloud/description: Allows Ingress from pods with label 'app=aggregate-prometheus'
+  creationTimestamp: null
+  name: allow-from-aggregate-prometheus
```
Are we sure that we need to allow ingress from `garden/aggregate-prometheus` but not from `garden/prometheus`?

To my knowledge `aggregate-prometheus` is only scraping other prometheus instances. See https://github.com/gardener/gardener/blob/a2fb2819758ad0687b2589845d7d2e43d5fa3ce2/docs/monitoring/README.md#L21-L26

FYI I deployed this change in my setup and some of the istio dashboards stopped showing data:
![Screenshot 2022-10-18 at 13 13 33](https://user-images.githubusercontent.com/9372594/196408211-512c660f-80f3-44e2-810c-cdded169495c.png)

axel7born replied:
Thanks looking at the dashboards!
It is really `aggregate-prometheus` that is scraping the data.
However, the issue was that the wrong label has been used as pod selector. 
Now, metrics are showing-up.

ialidzhikov replied:
Hm, I am not quite sure. I started my local setup against the latest revision from this PR and its the same, some of the dashboard stopped showing data:

![Screenshot 2022-10-19 at 12 29 56](https://user-images.githubusercontent.com/9372594/196653875-b725174c-f84c-4456-a3ce-1397127b0952.png)

ialidzhikov replied:
I think the reason for this is that the `allow-from-aggregate-prometheus` netpol has:

```yaml
    ports:
    - port: metrics
      protocol: TCP
```

while there is no such port with the given name in the istiod Pods:
```yaml
          ports:
          - containerPort: 15012
            protocol: TCP
          - containerPort: 10250
            protocol: TCP
          - containerPort: 8080
            protocol: TCP
```

axel7born replied:
Thanks! I fixed it.

ialidzhikov reviewed (commented):
I have only 1 inline comment. Otherwise lgtm.

axel7born committed:
Fix istio-system monitoring.

Add constant for "vpn-shoot"
Add description to the "allow-to-shoot-vpn-seed-server" policy.

ScheererJ commented on pkg/operation/botanist/component/vpnshoot/vpnshoot.go:
```diff
@@ -50,7 +50,7 @@ import (
 
 const (
-	labelValue = "vpn-shoot"
+	LabelValue = "vpn-shoot"
```
Please add a comment to make the verification step happy.

ScheererJ reviewed (approved):
Overall looks good, but please add the missing comment for the exported constant.

/lgtm

rfranzke reviewed (commented):
cc @plkokanov, this seems to be close to be merged. Can we add it to the v1.58 milestone?

plkokanov commented:
/milestone v1.58

ScheererJ reviewed (approved):
/lgtm
/approve

axel7born committed:
Fix port in netpol "allow-to-shoot-vpn-seed-server"

ScheererJ reviewed (approved):
/lgtm

ialidzhikov commented on pkg/operation/botanist/component/istio/istiod_test.go:
```diff
@@ -1937,7 +1937,7 @@ spec:
           role: monitoring
     ports:
-    - port: metrics
+    - port: 15014
```
Cosmetic suggestion to keep it idiomatic with the other prometheus netpols:
Lets preserve the named port here and add the following port to the istiod Pod template:
```yaml
          - containerPort: 15014
            name: metrics
            protocol: TCP
```

axel7born replied:
I'm a bit hesitant here, as we have the deployment as chart that we have to update from time to time with upstream istio charts. 
Should we really add this additional diff?
@ScheererJ, what do you think?

ialidzhikov replied:
Ah, okay. if the upstream manifest does not have the port, then let's keep it as it is. 👍

ialidzhikov commented:
@axel7born , can you run `make format` to fix the failing `pull-gardener-unit` test?

axel7born committed:
Fix format.

ialidzhikov reviewed (approved):
/lgtm
/approve
